keps/sig-api-machinery/2309-idl/README.md

```diff
@@ -0,0 +1,636 @@
+  OpenAPI tooling.
+
+* **Describes invalid types**: OpenAPI can easily describe constructs
+  which are either invalid (e.g. certain combinations of oneof under
+  structural rules) or not recommended (e.g. complex maps, floats).  Many
+  of these are tools that users reach for instinctively, and as such
+  OpenAPI guides users down the wrong path (see goal #1)
+
+* **Not a particularly good code generation source**: for similar reasons
+  to “hard to write by hand”, OpenAPI does not actually make
+  a particularly good code generation source for type-safe APIs.  For
+  instance:
+
+  * Existing tooling is pretty good at generating “naive” types from
+    OpenAPI, it’s hard to extract meaning on more complex types (e.g.
+    unions, enums).
+
+  * It’s currently impossible to identify stringly-typed types more
+    robustly in CRD OpenAPI due to type erasure.
+
+    <details>
+    <summary>stringly typed values...</summary>
+
+    ... are values whose concrete type is string, but which have
+    structured data inside the string.  Examples in k8s include Quantity,
+    Time, and Duration.
+
+    </details>
+
+  * Extensions could improve these cases, but that means:
+
+    * maintaining custom tooling to work with those custom extensions (not
+      that different from maintaining tooling that ingests IDL IR)
+
+    * Potentially waiting for API changes & API server version support
+      (often up to 1y after release), unless you want to maintain an extra
+      translation layer for older API servers.
+
+### Proto
+
+* Strongly tied to wire format: proto is strongly tied to our proto wire
+  format, which could cause many of the same issues as we see today with
+  the literal JSON representation being mirrored in Go code.
+
+  For instance, one might think to represent a union in proto as:
+
+   ```protobuf
+   message HasAUnion {
+     oneof {
+       VariantA a = 1;
+       VariantB b = 2;
+     }
+   }
+   ```
+
+  However, the current Kubernetes API guidelines require the use of
+  a tag.  While it would be possible to compile to OpenAPI that described
+  a tag, according to the way that our current proto serialization worked,
+  we’d also need to translate to a separate proto IDL file that also
+  included a tag for proto.  This seems confusing at best -- current users
+  express confusion today over “which can’t I just embed this arbitrary
+  struct into my pseudo-IDL, it’s all just Go code anyway”.
+
+  This also restricts how we change the IDL in the future -- unless we
+  have a separate pass to compile to proto, we’re restricted by not
+  breaking serialization compatibility.
+
+* **Describes invalid types**: for instance, proto has many integer types,
+  yet only a few are supported by the kubernetes type system, and we have
+  specific requirements around their proto representation (see “strongly
+  tied to wire format”)
+
+* **Mismatched semantics between proto keywords and Kubernetes**: for
```

Comment:
Note that we don't support proto3 today, but a future proto syntax across versions is not required to be wire compatible with our existing proto syntax.  Any such IDL we use in k/k probably needs to be convertible to protobuf, otherwise we will end up having to create a different conversion for proto3.  Hence why it may be problematic to assume k/k could pick up any new IDL.

Reply:
Can you reword this?  I'm having trouble parsing what you're saying after "but a future proto syntax across versions is not required to be wire compatible with our existing proto syntax".

In case this answers your question though: it's intended that this be able to continue to support our proto pipeline.  One of the nice things about having a separate IDL is that we can cleanly define semantics when converting to different forms: "this is what such-and-such a thing means in JSON, this is what we'd do when we generate the proto definitions & Go structs, etc".